azkaban-common/src/main/java/azkaban/executor/container/KubernetesContainerizedImpl.java

```diff
@@ -937,6 +1077,24 @@ V1Pod createPodFromMetadataAndSpec(final V1ObjectMeta podMetadata, final V1PodSp
   private void createPod(final int executionId) throws ExecutorManagerException {
     // Fetch execution flow from execution Id.
     final ExecutableFlow flow = this.executorLoader.fetchExecutableFlow(executionId);
+    FlowResourceRecommendation flowResourceRecommendation;
+    // Fetch flow resource recommendation for the given execution flow
+    Project project = this.projectManager.getProject(flow.getProjectId());
+    synchronized (project) {
+      final Map<String, FlowResourceRecommendation> flowResourceRecommendationMap =
+          project.getFlowResourceRecommendationMap();
+      flowResourceRecommendation = flowResourceRecommendationMap.get(flow.getFlowId());
+      // Migration: for all old flows, flow resource recommendation hasn't been generated yet in the DB
+      if (flowResourceRecommendation == null) {
+        flowResourceRecommendation =
+            this.projectManager.createFlowResourceRecommendation(flow.getProjectId(),
+                flow.getFlowId());
+
+        flowResourceRecommendationMap.put(flow.getFlowId(), flowResourceRecommendation);
+      }
+    }
+
+
```

Comment:
I think this can be simplified with two changes:
1. change the "flowResourceRecommendations" in Project class to a ConcurrentHashMap
2. replace this code piece with flowResourceRecommendations.computeIfAbsent(flow.getFlowId(), projectManger#create...)

What do you think? Is locking on the project instance necessary? Please don't hesitate to correct me if I miss anything.

Reply:
That's an excellent point!
I will address it in next PR b/c it will affect other components that are unchanged from this PR